flytekit/core/promise.py

```diff
@@ -916,7 +916,17 @@ def create_and_link_node(
                             )
                         is_optional = True
             if not is_optional:
-                raise _user_exceptions.FlyteAssertion("Input was not specified for: {} of type {}".format(k, var.type))
+                from flytekit.core.base_task import Task
+
+                error_msg = f"Input {k} of type {var.type} was not specified for function {entity.name}"
+
+                if isinstance(entity, Task):
+                    error_msg += (
+                        ". Flyte workflow syntax is a domain-specific language (DSL) for building execution graphs which "
+                        "supports a subset of Python’s semantics. When calling tasks, all kwargs have to be provided."
+                    )
+
+                raise _user_exceptions.FlyteAssertion(error_msg)
```

Comment:
In the review, it would be good to think whether the if statement `if isinstance(entity, Task) and _default is not None:` is not covering cases where this error message should be added as well or does cover cases where it shouldn't be added.

I tested the following:

1.
```python
@task
def foo(bar: int = 1) -> int:
    return bar

@workflow
def wf() -> int:
    return foo()
```

```console
flytekit.exceptions.user.FlyteAssertion: Error encountered while executing 'wf':
  Input bar of type <FlyteLiteral simple: INTEGER> was not specified for function test.foo.
  Flyte workflow syntax is a domain-specific language (DSL) for building execution graphs which supports a subset of Python’s semantics. When calling tasks, all kwargs have to be provided.
```

2.
```python
@task
def foo(bar: int) -> int:  # No default value
    return bar

@workflow
def wf() -> int:
    return foo()
```

```console
flytekit.exceptions.user.FlyteAssertion: Error encountered while executing 'wf':
  Input bar of type <FlyteLiteral simple: INTEGER> was not specified for function test.foo
```

3.
```python
@task
def foo(bar: int) -> int:
    return bar

@workflow
def sub_wf(bar: int=1) -> int:
    return foo(bar=bar)

@workflow
def wf() -> int:
    return sub_wf()
```

No error, as expected.

4.
```python
@task
def foo(bar: int) -> int:
    return bar

@workflow
def sub_wf(bar: int) -> int:
    return foo(bar=bar)

@workflow
def wf() -> int:
    return sub_wf()
```

```console
flytekit.exceptions.user.FlyteAssertion: Error encountered while executing 'wf':
  Encountered error while executing workflow 'sub_wf':
  Input bar of type <FlyteLiteral simple: INTEGER> was not specified for function sub_wf
```

5. Now with dynamic
```python
@task
def foo(bar: int) -> int:
    return bar

@dynamic
def sub_wf(bar: int) -> int:  # no default value
    return foo(bar=bar)

@workflow
def wf() -> int:
    return sub_wf()
```

```console
flytekit.exceptions.user.FlyteAssertion: Error encountered while executing 'wf':
  Input bar of type <FlyteLiteral simple: INTEGER> was not specified for function test.sub_wf
```

6.
```python
@task
def foo(bar: int) -> int:
    return bar

@dynamic
def sub_wf(bar: int=1) -> int:
    return foo(bar=bar)

@workflow
def wf() -> int:
    return sub_wf()
```

```console
flytekit.exceptions.user.FlyteAssertion: Error encountered while executing 'wf':
  Input bar of type <FlyteLiteral simple: INTEGER> was not specified for function test.sub_wf. 
  Flyte workflow syntax is a domain-specific language (DSL) for building execution graphs which supports a subset of Python’s semantics. When calling tasks, all kwargs have to be provided.
```

Reply:
would it be better to show the python type in the error message? like
```
Input bar of type [int] was not specified for function test.
```
we can get the python type from `interface.inputs[k]`

Reply:
Good call, thanks!

https://github.com/flyteorg/flytekit/pull/1921/commits/c725eff9f197cf90b94ff7f07fc5096066407073